---
title: "S4classes"
author: "Mariachiara Bottini"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{S4classes}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>"
)
```

```{r setup}
library(S4classes)
```

# Introduction:
It provides a set of S4 classes for Protein-coding genes, lncRNA e microRNA.
Those classes are inherited from a virtual gene class.

# Creating a new Protein-coding gene object:
It inherit all the gene class information, plus proteinID & proteinSeq.
Creating **ProteinCodingGene** example:
```{r}
tp53 <- ProteinCodingGene(
    id = "ENSG00000141510",
    symbol = "TP53",
    name = "TP53 gene",
    description = "Tumor protein p53",
    chromosome = "chr17",
    start = 7668402, 
    end = 7687550,
    proteinID = "P04637",
    proteinSeq = "MAQEGTNLRVEYLDDLRNGNRK"
)

```

# Creating a new long-non-coding RNA object:
It should inherit all the gene class information, plus lncRNAID and lncRNASeq. 
Creating a new lncRNA gene **lncRNAGene** is easy as shown in this example:
```{r}
lncRNA_gene2 <- lncRNAGene(
    id = "ENSG00000276085",
    symbol = "NEAT1",
    name = "NEAT1 gene",
    description = "Nuclear Paraspeckle Assembly Transcript 1",
    chromosome = "chr11", 
    start = 65497758, 
    end = 65503100,
    lncRNAID = "ENST00000463720",
    RNASeq = "GUCCAUCGGGUCCUCGAGAGCUUAGAGUU"
)

```

# Creating a new microRNA object:
It should inherit all the gene class information, plus microRNAID and seedSeq.
Creating a new miRNA **microRNAGene** is easy as shown in this example:
```{r}
miRNA_gene <- microRNAGene(
    id = "ENSG00000277483",
    symbol = "MIR155",
    name = "miR-155 gene",
    description = "microRNA 155",
    chromosome = "chr21",
    start = 25575833, 
    end = 25575902,
    microRNAID = "MI0000681",
    seedSeq = "UUAAUGCUAAUUGUGAUAGGGGU"
)

```

# Symbol:
In order to get or set a new symbol gene; 
Such that the user does not need to directly access data slots.

## Getting Symbol:
Getting Symbol **GetSymbol** is easy as shown in this example:
```{r}
GetSymbol(miRNA_gene)
```

## Setting Symbol:
Setting Symbol **SetSymbol** is easy as shown in this example:
```{r}
SetSymbol(tp53, "B")
```

# End position:
in order to get or set a new end position for the gene.

## Getting End:
Getting Symbol **getEnd** is easy as shown in this example:
```{r}
getEnd(miRNA_gene)
```

## Setting End:
Setting Symbol **setSymbol** is easy as shown in this example:
```{r}
setEnd(lncRNA_gene2, 250)
```

# Start position:
in order to get or set a new start position for the gene.

## Getting Start:
Getting Symbol **getSymbol** is easy as shown in this example:
```{r}
getStart(miRNA_gene)
```

## Setting Start:
Setting Symbol **setSymbol** is easy as shown in this example:
```{r}
setStart(lncRNA_gene2, 50)
```

# Getting Lenght Product:
which returns the length of the product of a specific gene object.
Getting **lengthProduct** is as easy as shown in this example:
```{r}
lengthProduct(tp53)
```

```{r session-info}
sessionInfo()
```
